README.md
# serialize-json
A serialization algorithm for JSON


History.md

1.0.2 / 2017-09-22
==================

  * refactor: enhance encode/decode large array performance (#4)

1.0.1 / 2017-02-09
==================

  * fix: support encode object without prototype (#2)

1.0.0 / 2017-02-06
==================

 * feat: implement a serialize algorithm for JSON


test/index.test.js
'use strict';

const JSON = require('../lib');
const is = require('is-type-of');
const assert = require('assert');
const urlparse = require('url').parse;

describe('test/index.test.js', () => {

  it('should encode & decode boolean', () => {
    let buf = JSON.encode(true);
    assert(is.buffer(buf));
    assert(JSON.decode(buf) === true);
    buf = JSON.encode(false);
    assert(JSON.decode(buf) === false);
  });

  it('should encode & decode string', () => {
    let buf = JSON.encode('xxx_yyy_中文 +|^%');
    assert(is.buffer(buf));
    assert(JSON.decode(buf) === 'xxx_yyy_中文 +|^%');
    buf = JSON.encode('');
    assert(is.buffer(buf));
    assert(JSON.decode(buf) === '');
  });

  it('should encode & decode integer', () => {
    const buf = JSON.encode(123456);
    assert(is.buffer(buf));
    assert(JSON.decode(buf) === 123456);
  });

  it('should encode & decode float', () => {
    const buf = JSON.encode(123.456);
    assert(is.buffer(buf));
    assert(JSON.decode(buf) === 123.456);
  });

  it('should encode & decode null', () => {
    const buf = JSON.encode(null);
    assert(is.buffer(buf));
    assert(JSON.decode(buf) === null);
  });

  it('should encode & decode undefined', () => {
    const buf = JSON.encode(undefined);
    assert(is.buffer(buf));
    assert(JSON.decode(buf) === undefined);
  });

  it('should encode & decode date', () => {
    const date = new Date();
    const buf = JSON.encode(date);
    assert(is.buffer(buf));
    assert.deepEqual(JSON.decode(buf), date);
  });

  it('should encode & decode buffer', () => {
    const buffer = new Buffer('hello world');
    const buf = JSON.encode(buffer);
    assert(is.buffer(buf));
    assert.deepEqual(JSON.decode(buf), buffer);
  });

  it('should not encode & decode function', () => {
    const buf = JSON.encode(function foo() {});
    assert(is.buffer(buf));
    assert(JSON.decode(buf) === null);
  });

  it('should encode & decode array', () => {
    let buf = JSON.encode([ 1, 2, 3, 4 ]);
    assert(is.buffer(buf));
    assert.deepEqual(JSON.decode(buf), [ 1, 2, 3, 4 ]);

    buf = JSON.encode([ '1', '2', 3, '4' ]);
    assert(is.buffer(buf));
    assert.deepEqual(JSON.decode(buf), [ '1', '2', 3, '4' ]);
  });

  it('should encode & decode error', () => {
    const err = new Error('mock error');
    err.code = 'success';
    const buf = JSON.encode(err);
    assert(is.buffer(buf));
    const ret = JSON.decode(buf);
    assert(is.error(ret));
    assert(ret.message === 'mock error');
    assert(ret.stack === err.stack);
    assert(ret.code === 'success');
    assert.deepEqual(ret, err);
  });

  it('should encode & decode object', () => {
    let json = {
      a: 'a',
      b: 123,
      c: 123.456,
      d: [ 1, 2, 3 ],
      e: true,
      f: null,
      g: undefined,
      h: new Date(),
      i: new Buffer('this is a buffer'),
      j: new Error('this is a error'),
    };
    let buf = JSON.encode(json);
    assert(is.buffer(buf));
    assert.deepEqual(JSON.decode(buf), json);

    json = {
      fn() {},
    };
    buf = JSON.encode(json);
    assert(is.buffer(buf));
    assert.deepEqual(JSON.decode(buf), {
      fn: null,
    });
  });

  it('should encode & encode object without prototype', () => {
    const url = urlparse('https://github.com?s=123', true);
    // url.query => { s: '123' } has no prototype
    let buf = JSON.encode(url);
    assert(is.buffer(buf));
    assert.deepEqual(JSON.decode(buf), url);

    const obj = Object.create(null, {
      a: {
        enumerable: true,
        value: '123',
      },
    });
    buf = JSON.encode(obj);
    assert(is.buffer(buf));
    assert.deepEqual(JSON.decode(buf), obj);
  });
});


lib/decoder.js
'use strict';

const debug = require('debug')('serialize-json#JSONDecoder');

const TOKEN_TRUE = -1;
const TOKEN_FALSE = -2;
const TOKEN_NULL = -3;
const TOKEN_EMPTY_STRING = -4;
const TOKEN_UNDEFINED = -5;

const REG_STR_REPLACER = /\+|%2B|%7C|%5E|%25/g;
const DECODER_REPLACER = {
  '+': ' ',
  '%2B': '+',
  '%7C': '|',
  '%5E': '^',
  '%25': '%',
};
const TOKEN_SET = new Set([ '|', '$', '@', '*', '#', ']' ]);

class JSONDecoder {
  constructor() {
    this.dictionary = [];
    this.tokens = [];
    this.tokensIndex = 0;
  }

  _decodeString(str) {
    return str.replace(REG_STR_REPLACER, a => DECODER_REPLACER[a]);
  }

  _decodeDate(str) {
    return new Date(this._base36To10(str));
  }

  _base36To10(num) {
    return parseInt(num, 36);
  }

  _unpack() {
    const token = this.tokens[this.tokensIndex];
    switch (token) {
      case '@': // array
      {
        debug('--> unpack array begin');
        const arr = [];
        const tokensLen = this.tokens.length;
        for (this.tokensIndex++; this.tokensIndex < tokensLen; this.tokensIndex++) {
          const token = this.tokens[this.tokensIndex];
          if (token === ']') {
            debug('--> unpack array end, %j', arr);
            return arr;
          }
          arr.push(this._unpack());
        }
        return arr;
      }
      case '$': // object
      {
        debug('--> unpack plain object begin');
        const obj = {};
        const tokensLen = this.tokens.length;
        for (this.tokensIndex++; this.tokensIndex < tokensLen; this.tokensIndex++) {
          const token = this.tokens[this.tokensIndex];
          if (token === ']') {
            debug('--> unpack plain object end, %j', obj);
            return obj;
          }
          const key = this._unpack();
          this.tokensIndex++;
          obj[key] = this._unpack();
        }
        return obj;
      }
      case '*': // buffer
      {
        debug('--> unpack buffer begin');
        const arr = [];
        const tokensLen = this.tokens.length;
        for (this.tokensIndex++; this.tokensIndex < tokensLen; this.tokensIndex++) {
          const token = this.tokens[this.tokensIndex];
          if (token === ']') {
            debug('--> unpack buffer end, %j', arr);
            return new Buffer(arr);
          }
          arr.push(this._unpack());
        }
        return new Buffer(arr);
      }
      case '#': // error
      {
        debug('--> unpack error begin');
        const obj = {};
        const tokensLen = this.tokens.length;
        for (this.tokensIndex++; this.tokensIndex < tokensLen; this.tokensIndex++) {
          const token = this.tokens[this.tokensIndex];
          if (token === ']') {
            const err = new Error(obj.message);
            Object.assign(err, obj);
            debug('--> unpack error end, %j', err);
            return err;
          }
          const key = this._unpack();
          this.tokensIndex++;
          obj[key] = this._unpack();
        }
        const err = new Error(obj.message);
        Object.assign(err, obj);
        return err;
      }
      case TOKEN_TRUE:
        return true;
      case TOKEN_FALSE:
        return false;
      case TOKEN_NULL:
        return null;
      case TOKEN_EMPTY_STRING:
        return '';
      case TOKEN_UNDEFINED:
        return undefined;
      default:
        return this.dictionary[token];
    }
  }

  decode(buf) {
    this.dictionary = [];
    this.tokens = [];
    this.tokensIndex = 0;

    const packed = buf.toString();
    const arr = packed.split('^');

    if (arr[0]) {
      const strArr = arr[0].split('|');
      for (const str of strArr) {
        this.dictionary.push(this._decodeString(str));
      }
    }

    if (arr[1]) {
      const intArr = arr[1].split('|');
      for (const int of intArr) {
        this.dictionary.push(this._base36To10(int));
      }
    }

    if (arr[2]) {
      const floatArr = arr[2].split('|');
      for (const float of floatArr) {
        this.dictionary.push(parseFloat(float));
      }
    }

    if (arr[3]) {
      const dateArr = arr[3].split('|');
      for (const date of dateArr) {
        this.dictionary.push(this._decodeDate(date));
      }
    }

    debug('decode packed json => %s, with dictionary %j', packed, this.dictionary);

    let tmp = '';
    for (let i = 0, len = arr[4].length; i < len; ++i) {
      const symbol = arr[4][i];
      if (TOKEN_SET.has(symbol)) {
        if (tmp) {
          this.tokens.push(this._base36To10(tmp));
          tmp = '';
        }
        if (symbol !== '|') {
          this.tokens.push(symbol);
        }
      } else {
        tmp += symbol;
      }
    }
    if (tmp) {
      this.tokens.push(this._base36To10(tmp));
    }
    return this._unpack();
  }
}

module.exports = JSONDecoder;


benchmark/decode.js
'use strict';

const Benchmark = require('benchmark');
const benchmarks = require('beautify-benchmark');
const suite = new Benchmark.Suite();

const JSONSerialize = require('..');

const json = {
  anyhost: true,
  application: 'demo-provider',
  dubbo: '2.0.0',
  generic: false,
  interface: 'com.alibaba.dubbo.demo.DemoService',
  loadbalance: 'roundrobin',
  methods: 'sayHello',
  owner: 'william',
  pid: 81281,
  side: 'provider',
  timestamp: 1481613276143,
};

const buf1 = new Buffer(JSON.stringify(json));
const buf2 = JSONSerialize.encode(json);

// add tests
suite
  .add('JSON.parse(buf1.toString())', function() {
    JSON.parse(buf1.toString());
  })
  .add('JSONSerialize.decode(buf2)', function() {
    JSONSerialize.decode(buf2);
  })
  .on('cycle', function(event) {
    benchmarks.add(event.target);
  })
  .on('start', function() {
    console.log('\n  node version: %s, date: %s\n  Starting...', process.version, Date());
  })
  .on('complete', function done() {
    benchmarks.log();
  })
  .run({ async: false });

// node version: v6.9.2, date: Mon Feb 06 2017 21:33:01 GMT+0800 (CST)
// Starting...
// 2 tests completed.

// JSON.parse(buf1.toString()) x 464,798 ops/sec ±1.62% (90 runs sampled)
// JSONSerialize.decode(buf2)  x  45,589 ops/sec ±1.21% (90 runs sampled)


benchmark/encode.js
'use strict';

const Benchmark = require('benchmark');
const benchmarks = require('beautify-benchmark');
const suite = new Benchmark.Suite();

const JSONSerialize = require('..');

const json = {
  anyhost: true,
  application: 'demo-provider',
  dubbo: '2.0.0',
  generic: false,
  interface: 'com.alibaba.dubbo.demo.DemoService',
  loadbalance: 'roundrobin',
  methods: 'sayHello',
  owner: 'william',
  pid: 81281,
  side: 'provider',
  timestamp: 1481613276143,
};

console.log(new Buffer(JSON.stringify(json)).length);
console.log(JSONSerialize.encode(json).length);

// add tests
suite
  .add('new Buffer(JSON.stringify(json))', function() {
    new Buffer(JSON.stringify(json));
  })
  .add('JSONSerialize.encode(json)', function() {
    JSONSerialize.encode(json);
  })
  .on('cycle', function(event) {
    benchmarks.add(event.target);
  })
  .on('start', function() {
    console.log('\n  node version: %s, date: %s\n  Starting...', process.version, Date());
  })
  .on('complete', function done() {
    benchmarks.log();
  })
  .run({ async: false });

// node version: v6.9.2, date: Mon Feb 06 2017 21:31:12 GMT+0800 (CST)
// Starting...
// 2 tests completed.

// new Buffer(JSON.stringify(json)) x 365,334 ops/sec ±1.44% (87 runs sampled)
// JSONSerialize.encode(json)       x  43,591 ops/sec ±1.23% (85 runs sampled)


benchmark/large_array.js
'use strict';

const Benchmark = require('benchmark');
const benchmarks = require('beautify-benchmark');
const suite = new Benchmark.Suite();

let count = 0;
const base = 900000000000000;
const userIds = [];
while (count < 50000) {
  userIds.push(String(base + count++));
}

console.log(userIds.slice(0, 10), userIds.slice(-10));

const JSONSerialize = require('..');

const buf_1 = new Buffer(JSON.stringify(userIds));
const buf_2 = JSONSerialize.encode(userIds);

// add tests
suite
  .add('new Buffer(JSON.stringify(json))', function() {
    new Buffer(JSON.stringify(userIds));
  })
  .add('JSONSerialize.encode(json)', function() {
    JSONSerialize.encode(userIds);
  })
  .add('JSON.parse(buf)', function() {
    JSON.parse(buf_1);
  })
  .add('JSONSerialize.decode(buf)', function() {
    JSONSerialize.decode(buf_2);
  })
  .on('cycle', function(event) {
    benchmarks.add(event.target);
  })
  .on('start', function() {
    console.log('\n  node version: %s, date: %s\n  Starting...', process.version, Date());
  })
  .on('complete', function done() {
    benchmarks.log();
  })
  .run({ async: false });

// node version: v8.4.0, date: Sat Sep 02 2017 08:58:29 GMT+0800 (CST)
// Starting...
// 4 tests completed.

// new Buffer(JSON.stringify(json)) x   255 ops/sec ±2.50% (74 runs sampled)
// JSONSerialize.encode(json)       x 28.50 ops/sec ±2.55% (50 runs sampled)
// JSON.parse(buf)                  x   201 ops/sec ±2.84% (70 runs sampled)
// JSONSerialize.decode(buf)        x 35.37 ops/sec ±4.61% (48 runs sampled)


benchmark/large_object.js
'use strict';

const Benchmark = require('benchmark');
const benchmarks = require('beautify-benchmark');
const suite = new Benchmark.Suite();

let count = 0;
const base = 900000000000000;
const obj = {};
while (count < 5000) {
  const userId = String(base + count++);
  obj[userId] = {
    date: new Date(),
    num: 10000000,
    bool: true,
    null: null,
    undefined: undefined,
    buf: new Buffer('this is a buffer'),
    err: new Error('this is a error'),
  };
}

const JSONSerialize = require('..');

const buf_1 = new Buffer(JSON.stringify(obj));
const buf_2 = JSONSerialize.encode(obj);

// add tests
suite
  .add('new Buffer(JSON.stringify(json))', function() {
    new Buffer(JSON.stringify(obj));
  })
  .add('JSONSerialize.encode(json)', function() {
    JSONSerialize.encode(obj);
  })
  .add('JSON.parse(buf)', function() {
    JSON.parse(buf_1);
  })
  .add('JSONSerialize.decode(buf)', function() {
    JSONSerialize.decode(buf_2);
  })
  .on('cycle', function(event) {
    benchmarks.add(event.target);
  })
  .on('start', function() {
    console.log('\n  node version: %s, date: %s\n  Starting...', process.version, Date());
  })
  .on('complete', function done() {
    benchmarks.log();
  })
  .run({ async: false });

// node version: v8.4.0, date: Sat Sep 02 2017 09:04:07 GMT+0800 (CST)
// Starting...
// 4 tests completed.

// new Buffer(JSON.stringify(json)) x 47.75 ops/sec ±2.27% (61 runs sampled)
// JSONSerialize.encode(json)       x  5.12 ops/sec ±4.66% (18 runs sampled)
// JSON.parse(buf)                  x 73.03 ops/sec ±5.54% (63 runs sampled)
// JSONSerialize.decode(buf)        x  5.81 ops/sec ±2.83% (19 runs sampled)


lib/encoder.js
'use strict';

const debug = require('debug')('serialize-json#JSONEncoder');
const is = require('is-type-of');
const utility = require('utility');

const REG_STR_REPLACER = /[\+ \|\^\%]/g;
const ENCODER_REPLACER = {
  ' ': '+',
  '+': '%2B',
  '|': '%7C',
  '^': '%5E',
  '%': '%25',
};

const TOKEN_TRUE = -1;
const TOKEN_FALSE = -2;
const TOKEN_NULL = -3;
const TOKEN_EMPTY_STRING = -4;
const TOKEN_UNDEFINED = -5;

class JSONEncoder {
  constructor() {
    this.dictionary = null;
  }

  encode(json) {
    this.dictionary = {
      strings: [],
      integers: [],
      floats: [],
      dates: [],
    };
    const ast = this._buildAst(json);

    let packed = this.dictionary.strings.join('|');
    packed += `^${this.dictionary.integers.join('|')}`;
    packed += `^${this.dictionary.floats.join('|')}`;
    packed += `^${this.dictionary.dates.join('|')}`;
    packed += `^${this._pack(ast)}`;

    debug('pack the json => %s', packed);
    return new Buffer(packed);
  }

  _pack(ast) {
    if (is.array(ast)) {
      let packed = ast.shift();
      for (const item of ast) {
        packed += `${this._pack(item)}|`;
      }
      return (packed[packed.length - 1] === '|' ? packed.slice(0, -1) : packed) + ']';
    }
    const type = ast.type;
    const index = ast.index;
    const dictionary = this.dictionary;
    const strLen = dictionary.strings.length;
    const intLen = dictionary.integers.length;
    const floatLen = dictionary.floats.length;

    switch (type) {
      case 'string':
        return this._base10To36(index);
      case 'integer':
        return this._base10To36(strLen + index);
      case 'float':
        return this._base10To36(strLen + intLen + index);
      case 'date':
        return this._base10To36(strLen + intLen + floatLen + index);
      default:
        return this._base10To36(index);
    }
  }

  _encodeString(str) {
    return str.replace(REG_STR_REPLACER, a => ENCODER_REPLACER[a]);
  }

  _base10To36(num) {
    return num.toString(36).toUpperCase();
  }

  _dateTo36(date) {
    return this._base10To36(date.getTime());
  }

  _buildStringAst(str) {
    const dictionary = this.dictionary;
    if (str === '') {
      return {
        type: 'empty',
        index: TOKEN_EMPTY_STRING,
      };
    }
    const data = this._encodeString(str);
    return {
      type: 'string',
      index: dictionary.strings.push(data) - 1,
    };
  }

  _buildNumberAst(num) {
    const dictionary = this.dictionary;
    // integer
    if (num % 1 === 0) {
      const data = this._base10To36(num);
      return {
        type: 'integer',
        index: dictionary.integers.push(data) - 1,
      };
    }
    // float
    return {
      type: 'float',
      index: dictionary.floats.push(num) - 1,
    };
  }

  _buildObjectAst(obj) {
    if (obj === null) {
      return {
        type: 'null',
        index: TOKEN_NULL,
      };
    }
    if (is.date(obj)) {
      const dictionary = this.dictionary;
      const data = this._dateTo36(obj);
      return {
        type: 'date',
        index: dictionary.dates.push(data) - 1,
      };
    }
    let ast;
    if (is.array(obj)) {
      ast = [ '@' ];
      for (const item of obj) {
        ast.push(this._buildAst(item));
      }
      return ast;
    }
    if (is.buffer(obj)) {
      ast = [ '*' ];
      for (const item of obj.values()) {
        ast.push(this._buildAst(item));
      }
      return ast;
    }
    if (is.error(obj)) {
      ast = [ '#' ];
      ast.push(this._buildAst('message'));
      ast.push(this._buildAst(obj.message));
      ast.push(this._buildAst('stack'));
      ast.push(this._buildAst(obj.stack));
    } else {
      ast = [ '$' ];
    }
    for (const key in obj) {
      // support object without prototype, like: Object.create(null)
      if (!utility.has(obj, key)) {
        continue;
      }
      ast.push(this._buildAst(key));
      ast.push(this._buildAst(obj[key]));
    }
    return ast;
  }

  _buildAst(item) {
    const type = typeof item;
    debug('calling buildAst with type: %s and data: %j', type, item);

    switch (type) {
      case 'string':
        return this._buildStringAst(item);
      case 'number':
        return this._buildNumberAst(item);
      case 'boolean':
        return {
          type: 'boolean',
          index: item ? TOKEN_TRUE : TOKEN_FALSE,
        };
      case 'undefined':
        return {
          type: 'undefined',
          index: TOKEN_UNDEFINED,
        };
      case 'object':
        return this._buildObjectAst(item);
      default:
        debug('unsupported type: %s, return null', type);
        return {
          type: 'null',
          index: TOKEN_NULL,
        };
    }
  }
}

module.exports = JSONEncoder;
